R/utils-hive-partitions.R

```diff
@@ -0,0 +1,175 @@
+#' - May have an empty value (interpreted as `NA` in most Hive/Arrow contexts)
+#'
+#' In strict mode, the function validates that all `key=value` segments are well-formed
+#' and will abort if any are not.
+#'
+#' @seealso [extract_hive_partitions()] to extract key-value pairs from Hive-style paths.
+#'
+#' @examples
+#' is_hive_partitioned_path("data/country=US/year=2024/file.parquet")
+#' is_hive_partitioned_path("data/country=/year=2024/", strict = TRUE)
+#' # is_hive_partitioned_path("data/=US/year=2024/", strict = TRUE) # This will error
+#'
+#' @export
+is_hive_partitioned_path <- function(path, strict = FALSE) {
+  checkmate::assert_character(path, len = 1)
+  # Split into path segments
+  parts <- fs::path_split(path)[[1]]
+
+  partition_like <- parts[grepl("=", parts)]
+  # Match valid Hive-style partitions
+  # Accepts: key=value or key= (missing value → NA)
+  valid <- grepl("^[^/]+=[^/]*$", partition_like)
+
+  if (strict && any(!valid)) {
+    cli::cli_abort(c(
+      "Invalid Hive-style partition segments detected in {.path {path}}:",
+      "x" = paste0("{.val ", partition_like[!valid], "}")
+    ))
+  }
+
+  any(valid)
+}
+#' Extract Hive-style partition key-value pairs from a path
+#'
+#' Given a filesystem path, this function extracts Hive-style partition
+#' key-value pairs (i.e., path components formatted as `key=value`). It supports
+#' decoding URL-encoded values (e.g., `"wk%20flu"` → `"wk flu"`), and handles
+#' empty values (e.g., `"key="`) as `NA`, consistent with Hive and Arrow semantics.
+#'
+#' If `strict = TRUE`, the function will abort with a detailed error message
+#' if any malformed partition-like segments are found.
+#'
+#' @param path A character string of length 1: the path to a file or directory.
+#' @param strict Logical. If `TRUE`, invalid partition segments (e.g., `=value`, or just `=`)
+#'   will trigger an error. If `FALSE`, only valid `key=value` components are returned.
+#'
+#' @return A named character vector where the names are partition keys and the values
+#'   are decoded values. Returns `NULL` if no valid partitions are found.
+#'
+#' @examples
+#' extract_hive_partitions("data/country=US/year=2024/file.parquet")
+#' extract_hive_partitions("data/country=/year=2024/", strict = TRUE)
+#' # extract_hive_partitions("data/=US/year=2024/", strict = TRUE) # This will error
+#' @seealso [is_hive_partitioned_path()]
+#' @export
+extract_hive_partitions <- function(path, strict = FALSE) {
+  checkmate::assert_character(path, len = 1)
+  if (!is_hive_partitioned_path(path, strict = strict)) {
+    return(NULL)
+  }
+  # Split the path
+  split_path <- fs::path_split(path)[[1]]
+
+  # Keep only the parts that contain an '=' sign — candidate key=value pairs
+  kv_pairs <- split_path[grepl("=", split_path)]
+
+  # Extract key-value pairs safely
+  purrr::map(kv_pairs, function(pair) {
+    # Split on the first '='
+    parts <- utils::strcapture("^([^=]+)=(.*)$",
```

Comment:
I'm always so happy whenever I see this function being used in the wild 🤩 It's so useful and yet not many know about it

Reply:
Hehe! The first time I saw it was in one of your PRs a while back! I like it a lot now too!